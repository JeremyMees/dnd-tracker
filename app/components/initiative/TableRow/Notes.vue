<script setup lang="ts">
import { INITIATIVE_SHEET } from '~~/constants/provide-keys'

const props = defineProps<{ item: InitiativeSheetRow }>()

const { sheet, update } = validateInject(INITIATIVE_SHEET)

const note = ref<string>(props.item.note || '')

watchDebounced(note, () => {
  if (!sheet.value) return

  const index = getCurrentRowIndex(sheet.value, props.item.id)
  const rows = [...sheet.value.rows]

  if (index === -1 || !rows[index]) return

  rows[index] = {
    ...rows[index],
    note: note.value,
  }

  update({ rows })
}, { debounce: 500, maxWait: 1000 })
</script>

<template>
  <div class="min-w-[150px] h-full relative">
    <textarea
      v-model="note"
      class="border-background border rounded-lg p-1 outline-none text-sm bg-muted resize-vertical w-full h-full min-h-10 max-h-[200px]"
      maxlength="250"
    />
  </div>
</template>
