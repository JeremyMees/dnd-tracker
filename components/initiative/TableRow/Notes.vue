<script setup lang="ts">
const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
}>()

const note = ref<string>(props.item.note || '')

watchDebounced(note, () => {
  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = props.sheet.rows

  if (index === -1) return

  rows[index].note = note.value

  props.update({ rows })
}, { debounce: 500, maxWait: 1000 })
</script>

<template>
  <td class="min-w-[150px] relative">
    <textarea
      v-model="note"
      class="border-slate-700 border rounded-lg p-2 outline-none body-small bg-bg-light resize-vertical w-full h-10"
      maxlength="250"
    />
  </td>
</template>
