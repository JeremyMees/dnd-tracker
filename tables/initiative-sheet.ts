import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
