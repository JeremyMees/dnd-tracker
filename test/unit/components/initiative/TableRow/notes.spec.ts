import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach, afterEach } from 'vitest'
import Notes from '~/components/initiative/TableRow/Notes.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row notes', async () => {
  beforeEach(() => {
    mockUpdate.mockClear()
    vi.useFakeTimers()
  })

  afterEach(() => vi.useRealTimers())

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Notes, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should initialize with empty note when no note is provided', async () => {
    const component = await mountSuspended(Notes, {
      props: {
        ...props,
        item: { ...props.item, note: undefined },
      },
    })

    expect(component.find('textarea').element.value).toBe('')
  })

  it('Should initialize with existing note', async () => {
    const component = await mountSuspended(Notes, {
      props: {
        ...props,
        item: { ...props.item, note: 'Test note' },
      },
    })

    expect(component.find('textarea').element.value).toBe('Test note')
  })

  it('Should debounce note updates', async () => {
    const component = await mountSuspended(Notes, { props })
    const textarea = component.find('textarea')

    await textarea.setValue('New note')
    expect(mockUpdate).not.toHaveBeenCalled()

    vi.advanceTimersByTime(400)
    expect(mockUpdate).not.toHaveBeenCalled()

    vi.advanceTimersByTime(100)
    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          id: props.item.id,
          note: 'New note',
        }),
      ]),
    })
  })

  it('Should not update when sheet is undefined', async () => {
    const component = await mountSuspended(Notes, {
      props: {
        ...props,
        sheet: undefined,
      },
    })

    await component.find('textarea').setValue('New note')
    vi.advanceTimersByTime(500)

    expect(mockUpdate).not.toHaveBeenCalled()
  })
})
