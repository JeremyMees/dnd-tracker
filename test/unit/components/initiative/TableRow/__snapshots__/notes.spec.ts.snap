// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row notes > Should match snapshot 1`] = `"<div class="min-w-[150px] h-full relative"><textarea class="border-background border rounded-lg p-1 outline-none text-sm bg-muted resize-vertical w-full h-full min-h-10 max-h-[200px]" maxlength="250"></textarea></div>"`;
